### PR TITLE
TST: Update list of modules for import cycle checks

### PR DESCRIPTION
**Pull request by scottshambaugh**
#### What does this implement/fix?
The list of modules in the import cycles check was missing a few new modules (including _spatial.transform_, _stats.qmc_, _stats.sampling_, _datasets_, and maybe one or two others). Updated to import the list from [test_public_api](https://github.com/scipy/scipy/blob/main/scipy/_lib/tests/test_public_api.py), which should fail if a module is added to scipy but not to that list, so the problem of the import cycles list being accidentally unmaintained should no longer be an issue.

### CODE REVIEW

**Comment by scottshambaugh** [comment]
It looks like this might have caught an error:
```
=================================== FAILURES ===================================
___________________________ test_modules_importable ____________________________
[gw0] linux -- Python 3.8.10 /usr/bin/python3.8-dbg
../testenv/lib/python3.8/site-packages/scipy/_lib/tests/test_import_cycles.py:12: in test_modules_importable
    subprocess.check_call([sys.executable, '-c', cmd])
        cmd        = 'import scipy.signal.spectral'
        module     = 'scipy.signal.spectral'
/usr/lib/python3.8/subprocess.py:359: in check_call
    retcode = call(*popenargs, **kwargs)
        kwargs     = {}
        popenargs  = (['/usr/bin/python3.8-dbg', '-c', 'import scipy.signal.spectral'],)
/usr/lib/python3.8/subprocess.py:342: in call
    return p.wait(timeout=timeout)
        kwargs     = {}
        p          = <subprocess.Popen object at 0x7f94b41f1f50>
        popenargs  = (['/usr/bin/python3.8-dbg', '-c', 'import scipy.signal.spectral'],)
        timeout    = None
/usr/lib/python3.8/subprocess.py:1083: in wait
    return self._wait(timeout=timeout)
        self       = <subprocess.Popen object at 0x7f94b41f1f50>
        timeout    = None
/usr/lib/python3.8/subprocess.py:1806: in _wait
    (pid, sts) = self._try_wait(0)
        self       = <subprocess.Popen object at 0x7f94b41f1f50>
        timeout    = None
/usr/lib/python3.8/subprocess.py:1764: in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
E   Failed: Timeout >60.0s
        self       = <subprocess.Popen object at 0x7f94b41f1f50>
        wait_flags = 0
----------------------------- Captured stderr call -----------------------------
```

File: scipy/_lib/tests/test_import_cycles.py
Comment (tupui) [inline scipy/_lib/tests/test_import_cycles.py:10]: I am not sure here, I would just test the public API. IMO it's ok if we accidentally break things on the private side and would force people to move away from doing such things.

**Review by tupui — COMMENTED** [review]
Thanks for the PR @scottshambaugh. I think we need a few maintainers opinions there.

File: scipy/_lib/tests/test_import_cycles.py
Comment (j-bowhay) [reply scipy/_lib/tests/test_import_cycles.py:10]: I tested this locally, it does past but it is very slow. Adding the private modules makes this test take ~4x longer.

Without private modules
`python dev.py test --tests scipy._lib.tests.test_import_cycles  17.32s user 57.71s system 848% cpu 8.838 total`
With private modules
`python dev.py test --tests scipy._lib.tests.test_import_cycles  63.69s user 227.98s system 1008% cpu 28.916 total`

Imagine this could probably be made faster if all imports were checked in one subproccess call rather than one for each import?

Comment (scottshambaugh) [reply scipy/_lib/tests/test_import_cycles.py:10]: I think I understand what’s happening. This test really wants to be a multiprocessing pool, but for the sake of simplicity I think it might work to break it up into two separate tests for the public and non-public modules. Let me try that out.

Comment (tupui) [reply scipy/_lib/tests/test_import_cycles.py]: Again, not sure we want to test private modules at all.

Comment (scottshambaugh) [reply scipy/_lib/tests/test_import_cycles.py]: Up to the maintainers, though IMO even if those modules are not stable API they should still be stable from an import perspective.

Comment (tupui) [reply scipy/_lib/tests/test_import_cycles.py]: Everything is a maintainability tradeoff. If the test passes with all private modules, we could add it not, but I would consider this more as informative than require this to be green. A compromise could be to mark the test as `xfail` and a final decision would be taken at release time when we review the failed tests.

File: scipy/_lib/tests/test_import_cycles.py
Comment (tupui) [inline scipy/_lib/tests/test_import_cycles.py:16]: I don't think we would do this. At most we can mark it `xslow`, but as it is the test would add >60s to all CI runs which is not acceptable.

**Review by tupui — COMMENTED** [review]
Thank you @scottshambaugh for the updates.

Please don't add more commit yet to preserve CI. We need to make a decision first.

**Comment by rgommers** [comment]
> In the interest of test coverage, does it make sense to remove the timeout check, only test the list of `PUBLIC_MODULES` , and merge now?

That sounds good to me. There's a clear benefit there, and should be straightforward to merge then.

**Comment by scottshambaugh** [comment]
Sounds good! It looks like the CI failures here are unrelated.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @scottshambaugh!

I measured locally, and test duration went from 8.7 to 9.1 seconds, which is as expected.

**Pull request merged**
